subsys/ipc/ipc_service/backends/ipc_icmsg.c

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <string.h>
+#include <drivers/mbox.h>
+
+#include <ipc/ipc_service_backend.h>
+#include "ipc_icmsg.h"
+
+#define DT_DRV_COMPAT	zephyr_ipc_icmsg
+#define CB_BUF_SIZE	CONFIG_IPC_SERVICE_BACKEND_ICMSG_CB_BUF_SIZE
+
+static const uint8_t magic[] = {0x45, 0x6d, 0x31, 0x6c, 0x31, 0x4b,
+				0x30, 0x72, 0x6e, 0x33, 0x6c, 0x69, 0x34};
+BUILD_ASSERT(sizeof(magic) <= CB_BUF_SIZE);
+BUILD_ASSERT(CB_BUF_SIZE <= UINT8_MAX);
+
+struct backend_data_t {
+	/* Tx/Rx buffers. */
+	struct icmsg_buf *tx_ib;
+	struct icmsg_buf *rx_ib;
+
+	/* Backend ops for an endpoint. */
+	const struct ipc_service_cb *ops;
+
+	/* General */
+	struct k_work mbox_work;
+	atomic_t state;
+};
+
+struct backend_config_t {
+	uintptr_t tx_shm_addr;
+	uintptr_t rx_shm_addr;
+	size_t tx_shm_size;
+	size_t rx_shm_size;
+	struct mbox_channel mbox_tx;
+	struct mbox_channel mbox_rx;
+};
+
+
+static void mbox_callback_process(struct k_work *item)
+{
+	struct backend_data_t *dev_data = CONTAINER_OF(item, struct backend_data_t, mbox_work);
+	uint8_t cb_buffer[CB_BUF_SIZE] __aligned(4);
+
+	atomic_t state = atomic_get(&dev_data->state);
+	int len = icmsg_buf_read(dev_data->rx_ib, cb_buffer, CB_BUF_SIZE);
+
+	__ASSERT_NO_MSG(len <= CB_BUF_SIZE);
+
+	if (len == -EAGAIN) {
+		(void)k_work_submit(&dev_data->mbox_work);
```

Comment:
imo this should be avoided. If callback is called there should be pending data. We can easily turn error condition into continuous polling that is not detected.

Reply:
My idea was to react to at least try to recover from unexpected behavior. Do you suggest this should be removed and ASSERTED?

Reply:
This should not occur in any case. If it does however the best we can do is to retry.

The only case where this is a problem is when, for some reason the other core dies out before whole message is delivered (should not happen as we update index with corresponding len). If it does we will get into resubmitting loop. Maybe indeed the best thing to do, assuming this cannot happen is return and do nothing...